### PR TITLE
Feature/PPM-320 create topology task

### PR DESCRIPTION
**Pull request by morantr**
Create the Topology Task and move all related code from the backhaul manager to the task.

Passed certification tests:
https://gitlab.com/prpl-foundation/prplMesh/-/pipelines/174069281

### CODE REVIEW

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 1]
```diff
@@ -245,10 +245,6 @@ backhaul_manager::backhaul_manager(const config_file::sConfigSlave &config,
     m_sConfig.vendor            = config.vendor;
     m_sConfig.model             = config.model;
 
-    // Create a local copy on this process database instance. Will be removed on PPM-83 phase 5
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:248]: suggested commit (fixed typos inside, and line-length not checked so should be handled by you):

```
Cleanup preparative commit.
PR #1504 has added all the members required for the Topology Task to the AgentDB.
Since at that time the AgentDB was not really shared since the son_slaves were on a different
process than the backhaul manager, we had to create a local copy of everything which is
passed on on messages between the son_slaves and the backhaul manager.
Now that PR #1584 is merged and the database is shared, we can delete all of these local
copies:
1. Removed passing local_gw and local_controller flags.
2. Removed passing bridge_mac on backhaul connected notification, and setting a local
copy of it on the backhaul constructor.
3. Removed passing of vap list update to the backhaul manager.
4. Removed passing client associated and disconnected messages to the backhaul manager.
5. Removed dependency on backhaul manager member "slaves_sockets" for "freq_type" and
remove it from the backhaul manager.
6. Removed adding the radio nodes again on the backhaul manager (it is added on the
son_slave already).
7. Removed the local copy of a bunch of parameters on the BACKHAUL ENABLE handler, that we now
have access to through the DB.

PPM-320

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 2]
```diff
@@ -1518,7 +1514,11 @@ bool backhaul_manager::backhaul_fsm_wireless(bool &skip_select)
                     LOG(WARNING) << "Sta_hal of " << soc->sta_iface << " is null";
                     continue;
                 }
-                if (m_sConfig.backhaul_preferred_radio_band == soc->freq_type) {
+                auto radio = db->radio(soc->hostap_iface);
+                if (!radio) {
+                    continue;
+                }
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1379]:
```suggestion
                auto radio = db->radio(soc->hostap_iface);
                if (!radio) {
                    LOG(DEBUG) << "No radio found in DB for interface " << soc->hostap_iface;
                    continue;
                }
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 3]
```diff
@@ -1538,9 +1538,14 @@ bool backhaul_manager::backhaul_fsm_wireless(bool &skip_select)
                 continue;
             }
 
+            auto radio = db->radio(soc->hostap_iface);
+            if (!radio) {
+                continue;
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1402]:
```suggestion
                LOG(DEBUG) << "No radio found in DB for interface " << soc->hostap_iface;
                continue;
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 4]
```diff
@@ -1943,20 +1946,10 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<sRadioInfo>
...
-        soc->radio_mac = request->iface_mac();
-        // Create a local copy on this process database instance. Will be removed on PPM-83 phase 5
-        radio->front.iface_mac = soc->radio_mac;
-
-        soc->freq_type = request->frequency_band();
-        // Create a local copy on this process database instance. Will be removed on PPM-83 phase 5
-        radio->front.freq_type     = request->frequency_band();
+        soc->radio_mac             = request->iface_mac();
         soc->controller_discovered = false;
-        soc->max_bandwidth         = request->max_bandwidth();
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1954]: should this have also been removed?

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 5]
```diff
@@ -3539,7 +3461,11 @@ bool backhaul_manager::handle_1905_autoconfiguration_response(ieee1905_1::CmduMe
         }
         LOG(DEBUG) << "received autoconfiguration response for " << band_name << " band";
         for (auto soc : slaves_sockets) {
-            if (soc->freq_type == freq_type) {
+            auto radio = db->radio(soc->hostap_iface);
+            if (!radio) {
+                continue;
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2920]:
```suggestion
                LOG(DEBUG) << "No radio found in DB for interface " << soc->hostap_iface;
                continue;
```

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 6]
```diff
@@ -1849,33 +1830,19 @@ bool slave_thread::handle_cmdu_ap_manager_message(Socket *sd,
...
-        LOG(INFO) << "client disconnected sta_mac=" << client_mac;
+        auto &client_mac = notification_in->params().mac;
+        auto &bssid      = notification_in->params().bssid;
+        LOG(INFO) << "client disconnected sta_mac=" << client_mac << " from bssid=" << bssid;
 
         //notify master
         if (!master_socket) {
             LOG(DEBUG) << "Controller is not connected";
             return true;
         }
```
Comment (adam1985d) [inline agent/src/beerocks/slave/son_slave_thread.cpp:1813]:
```suggestion
        //notify master
        if (!master_socket) {
            LOG(DEBUG) << "Controller is not connected";
            return true;
        }
        
        auto &client_mac = notification_in->params().mac;
        auto &bssid      = notification_in->params().bssid;
        LOG(INFO) << "Client disconnected sta_mac=" << client_mac << " from bssid=" << bssid;

```

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 7]
```diff
@@ -2036,37 +2003,29 @@ bool slave_thread::handle_cmdu_ap_manager_message(Socket *sd,
             return false;
         }
         LOG(TRACE) << "received ACTION_APMANAGER_CLIENT_ASSOCIATED_NOTIFICATION";
-        std::string client_mac = tlvf::mac_to_string(notification_in->mac());
-        LOG(INFO) << "client associated sta_mac=" << client_mac;
+        auto &client_mac = notification_in->mac();
+        auto &bssid      = notification_in->bssid();
+        LOG(INFO) << "client associated sta_mac=" << client_mac << " to bssid=" << bssid;
```
Comment (adam1985d) [inline agent/src/beerocks/slave/son_slave_thread.cpp]:
```suggestion
        LOG(INFO) << "Client associated sta_mac=" << client_mac << " to bssid=" << bssid;
```

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 8]
```diff
@@ -3123,27 +3134,30 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
...
         bool error = false;
         auto db    = AgentDB::get();
-        if (!db->ethernet.iface_name.empty()) {
-            if (config.backhaul_wire_iface_type == beerocks::IFACE_TYPE_UNSUPPORTED) {
-                LOG(DEBUG) << "backhaul_wire_iface_type is UNSUPPORTED";
-                platform_notify_error(
-                    bpl::eErrorCode::CONFIG_BACKHAUL_WIRED_INTERFACE_IS_UNSUPPORTED, "");
-                error = true;
+
+        if (!db->device_conf.local_gw) {
```
Comment (adam1985d) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: why is this required?
I ask since you clear the `db->device_conf.local_gw` in the `beerocks_message::ACTION_PLATFORM_SON_SLAVE_REGISTER_RESPONSE:` that happens before the enable.

either way, it looks like the entire functionality here is relevant only when not local_gw

so why not:
```
if (db->device_conf.local_gw) {
    LOG(TRACE) << "goto STATE_SEND_BACKHAUL_MANAGER_ENABLE";
    slave_state = STATE_SEND_BACKHAUL_MANAGER_ENABLE;
    break;
}
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 9]
```diff
@@ -1082,6 +1082,14 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
         for (auto soc : slaves_sockets) {
             std::string iface = soc->sta_iface;
 
+            auto radio = db->radio(soc->sta_iface);
+            if (!radio) {
+                LOG(DEBUG) << "Radio of iface " << soc->sta_iface << " does not exist on the db";
+                continue;
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:990]: if you continue here it means the socket HALs are not freed - is this what you intended?

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 10]
```diff
@@ -1082,6 +1082,14 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
         for (auto soc : slaves_sockets) {
             std::string iface = soc->sta_iface;
 
+            auto radio = db->radio(soc->sta_iface);
+            if (!radio) {
+                LOG(DEBUG) << "Radio of iface " << soc->sta_iface << " does not exist on the db";
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:989]:
```suggestion
                LOG(DEBUG) << "Radio of iface " << soc->sta_iface << " does not exist in the db";
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 11]
```diff
@@ -1436,6 +1444,15 @@ bool backhaul_manager::backhaul_fsm_wireless(bool &skip_select)
                 //     break;
                 // }
 
+                auto radio = db->radio(soc->sta_iface);
+                if (!radio) {
+                    LOG(DEBUG) << "Radio of iface " << soc->sta_iface
+                               << " does not exist on the db";
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1292]:
```suggestion
                               << " does not exist in the db";
```

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 12]
```diff
@@ -154,6 +154,16 @@ void slave_thread::slave_reset()
     is_backhaul_manager   = false;
     detach_on_conf_change = false;
 
+    auto db = AgentDB::get();
+
+    auto radio = db->radio(m_fronthaul_iface);
+    if (!radio) {
+        LOG(DEBUG) << "Radio of iface " << m_fronthaul_iface << " does not exist on the db";
+        return;
+    }
```
Comment (adam1985d) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: if you really wish to perform a `return` here and not exit via the "slave_state" like before, please provide a bit more information why the rest of the operations in the function shouldn't be performed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 13]
```diff
@@ -34,6 +34,7 @@
 #include "../helpers/media_type.h"
 #include "../link_metrics/ieee802_11_link_metrics_collector.h"
 #include "../link_metrics/ieee802_3_link_metrics_collector.h"
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:36]: commit message doesn't meet the contribution.md line-length limit.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 14]
```diff
@@ -186,6 +187,12 @@ backhaul_manager::backhaul_manager(const config_file::sConfigSlave &config,
 
     m_eFSMState = EState::INIT;
     set_select_timeout(SELECT_TIMEOUT_MSC);
+
+    auto topology_task = std::make_shared<TopologyTask>(*this, cmdu_tx);
+    if (!topology_task) {
+        LOG(ERROR) << "failed to allocate Topology Task!";
+    }
+    m_task_pool.add_task(topology_task);
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:195]: this should have been added separately to the task skeleton, after you really have a reason to start it
(nitpick)

File: agent/src/beerocks/slave/tasks/topology_task.cpp [hunk 15]
```diff
@@ -35,3 +35,15 @@ TopologyTask::TopologyTask(backhaul_manager &btl_ctx, ieee1905_1::CmduMessageTx
     : Task(eTaskType::TOPOLOGY), m_btl_ctx(btl_ctx), m_cmdu_tx(cmdu_tx)
 {
 }
+
```
Comment (adam1985d) [inline agent/src/beerocks/slave/tasks/topology_task.cpp:38]: excessive spacing in commit message after `Currently,` and `The` should have been `the`

File: agent/src/beerocks/slave/tasks/topology_task.h [hunk 16]
```diff
@@ -24,6 +24,9 @@ class TopologyTask : public Task {
     ~TopologyTask() {}
 
 private:
+    /* Helpers functions */
```
Comment (adam1985d) [inline agent/src/beerocks/slave/tasks/topology_task.h]:
```suggestion
    /* Helper functions */
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 17]
```diff
@@ -1091,66 +1091,6 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
     return (true);
 } // namespace beerocks
 
-bool backhaul_manager::send_1905_topology_discovery_message()
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1145]: commit message type `remeaned` should be renamed - and in general, it seems the commit messages in this PR need -rereview and alignment to contribuition.md guidelines

File: agent/src/beerocks/slave/tasks/topology_task.cpp [hunk 18]
```diff
@@ -36,6 +36,52 @@ TopologyTask::TopologyTask(backhaul_manager &btl_ctx, ieee1905_1::CmduMessageTx
...
+                std::chrono::seconds(DISCOVERY_NEIGHBOUR_REMOVAL_TIMEOUT_SEC)) {
+                auto &device_al_mac = it->first;
+                LOG(INFO) << "Removed 1905.1 device " << device_al_mac << " from neighbors list";
+                it                     = neighbors_on_local_iface.erase(it);
+                neighbors_list_changed = true;
+                continue;
+            }
+            it++;
+        }
+    }
```
Comment (adam1985d) [inline agent/src/beerocks/slave/tasks/topology_task.cpp:78]: this code screams "please encapsulate me by a private function with a meaningful name!"

File: agent/src/beerocks/slave/tasks/topology_task.cpp [hunk 19]
```diff
@@ -83,6 +83,104 @@ void TopologyTask::work()
...
+bool TopologyTask::handle_cmdu(ieee1905_1::CmduMessageRx &cmdu_rx, const sMacAddr &src_mac,
+                               std::shared_ptr<beerocks_header> beerocks_header)
+{
+    switch (cmdu_rx.getMessageType()) {
+    case ieee1905_1::eMessageType::TOPOLOGY_DISCOVERY_MESSAGE: {
+        handle_topology_discovery(cmdu_rx, src_mac);
+        break;
+    }
+    default: {
+        // Message was not handled, therfore return false.
```
Comment (adam1985d) [inline agent/src/beerocks/slave/tasks/topology_task.cpp]:
```suggestion
        // Message was not handled, therefore return false.
```

File: agent/src/beerocks/slave/tasks/topology_task.cpp [hunk 20]
```diff
@@ -83,6 +83,104 @@ void TopologyTask::work()
...
+    // Check if it is a new device so if it does, we will send a Topology Notification.
+    bool new_device = false;
+    for (auto &neighbors_on_local_iface_entry : db->neighbor_devices) {
+        auto &neighbors_on_local_iface = neighbors_on_local_iface_entry.second;
+        new_device =
+            neighbors_on_local_iface.find(tlvAlMac->mac()) == neighbors_on_local_iface.end();
+        if (new_device) {
+            break;
+        }
+    }
```
Comment (adam1985d) [inline agent/src/beerocks/slave/tasks/topology_task.cpp:180]: I know this code is ported as-is and maybe it is just me - but I think there must be a clearer way to implement this check.
one example would be:
```
if (neighbors_on_local_iface.find(tlvAlMac->mac()) == neighbors_on_local_iface.end()) {
    new_device = true;
    break;
}
```
a better way is a helper function to check "is_device_in_neigbors_map" or similar and call it here

File: agent/src/beerocks/slave/tasks/topology_task.h [hunk 21]
```diff
@@ -25,7 +25,20 @@ class TopologyTask : public Task {
...
 private:
+    /* 1905.1 message handlers: */
+
+    /**
+    * @brief Handles 1905 Topology Discovery message.
+    * 
+    * @param[in] cmdu_rx Received CMDU.
+    * @param[in] src_mac MAC address of the message sender.
+    */
+    void handle_topology_discovery(ieee1905_1::CmduMessageRx &cmdu_rx, const sMacAddr &src_mac);
```
Comment (adam1985d) [inline agent/src/beerocks/slave/tasks/topology_task.h:47]: 👍 
I liked how you reordered the function parameters since this API didn't align with the rest of similar APIs (meaning Message precedes the MAC and not the other way around and it was)

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 22]
```diff
@@ -366,14 +366,8 @@ bool backhaul_manager::socket_disconnected(Socket *sd)
 
             if (m_eFSMState >= EState::CONNECT_TO_MASTER) {
                 LOG(INFO) << "Sending topology notification on son_slave disconnect";
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:368]: I am not sure if the log print is used by tests for verification (aka Telemetry) - but if not should be changed to what is actually done here and move the print of sending topology notification inside.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 23]
```diff
@@ -1816,14 +1810,8 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<sRadioInfo>
 
         if (m_eFSMState >= EState::CONNECT_TO_MASTER) {
             LOG(INFO) << "Sending topology notification on reconnected son_slave";
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1821]: same as the above comment regarding the log print

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 24]
```diff
@@ -2112,6 +2112,11 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
     case ieee1905_1::eMessageType::BACKHAUL_STEERING_REQUEST_MESSAGE: {
         return handle_backhaul_steering_request(cmdu_rx, src_mac);
     }
+    case ieee1905_1::eMessageType::VENDOR_SPECIFIC_MESSAGE: {
+        // We should not handle vendor specific messages here, return false so the message will
+        // be forwarded and will not pass to the task_pool.
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]:
```suggestion
        // be forwarded and will not be passed to the task_pool.
```

File: agent/src/beerocks/slave/tasks/topology_task.cpp [hunk 25]
```diff
@@ -510,13 +513,49 @@ void TopologyTask::handle_topology_query(ieee1905_1::CmduMessageRx &cmdu_rx,
...
+                                          const sMacAddr &src_mac,
+                                          std::shared_ptr<beerocks_header> beerocks_header)
+{
+    if (!beerocks_header) {
+        LOG(ERROR) << "beerocks_header is nullptr";
+        return false;
+    }
+
+    // Since currently we handle only action_ops of action type "ACTION_BACKHAUL", use a single
+    // a single switch-case on "ACTION_BACKHAUL" only.
```
Comment (adam1985d) [inline agent/src/beerocks/slave/tasks/topology_task.cpp]:
```suggestion
    // Since currently we handle only action_ops of action type "ACTION_BACKHAUL", use
    // a single switch-case on "ACTION_BACKHAUL" only.
```
this looked like the reading tests you get on social media :-)

File: agent/src/beerocks/slave/tasks/topology_task.cpp [hunk 26]
```diff
@@ -207,14 +207,152 @@ void TopologyTask::handle_topology_query(ieee1905_1::CmduMessageRx &cmdu_rx,
         return;
     }
 
+    if (!add_device_information_tlv()) {
```
Comment (adam1985d) [inline agent/src/beerocks/slave/tasks/topology_task.cpp:218]: you should have split the "move to functions" and the changes in the implementation to 2 separate commits - this is really hard to review

File: ci/cppcheck/cppcheck_existing_issues.txt [hunk 27]
```diff
@@ -7,8 +7,6 @@ agent/src/beerocks/fronthaul_manager/monitor/monitor_thread.cpp: style: Variable
 agent/src/beerocks/fronthaul_manager/monitor/monitor_thread.cpp: style: Variable 'iface' is assigned a value that is never used. [unreadVariable]        std::string iface = radio_node->get_iface();
 agent/src/beerocks/fronthaul_manager/monitor/rdkb/monitor_rdkb_hal.cpp: style: Consecutive return, break, continue, goto or throw statements are unnecessary. [duplicateBreak]        break;
 agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp: style: Local variable 'iface_hal' shadows outer variable [shadowVariable]                            auto iface_hal = get_wireless_hal(sta_iface);
-agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp: style: Local variable 'local_interface_name' shadows outer variable [shadowVariable]        std::string local_interface_name = soc->hostap_iface;
-agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp: style: Local variable 'supportedServiceTuple' shadows outer variable [shadowVariable]        auto supportedServiceTuple = tlvSupportedService->supported_service_list(1);
```
Comment (adam1985d) [inline ci/cppcheck/cppcheck_existing_issues.txt:11]: this should be squashed to the commit that fixed it and shouldn't be as a separate commit

**Review by adam1985d — APPROVED** [review]
Please address the comments I raised before merging.
Since I think all are small changes - approving

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 1 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:248]: fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 4 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1954]: yes, because it is now on the database.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 2 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1379]: I prefer to not print anything here. The check is only for safety and there isn't a chance that the radio will no exist.
Either way, it is not something we would really care about it here.
Same for other places.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 3 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1402]: https://github.com/prplfoundation/prplMesh/pull/1576/commits/6a913a9fe165196dcdc31b2287c4b095dddd447d#r464056964

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 5 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2920]: https://github.com/prplfoundation/prplMesh/pull/1576/commits/6a913a9fe165196dcdc31b2287c4b095dddd447d#r464056964

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 6 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp:1813]: Actually the fact that the print comes first is important. I'll keep it as is.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 7 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: Fixed

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 8 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: I did it because I didn't want to fail here, anyway, I liked more what you have suggested so I'll change it to that.
Thanks.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 10 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:989]: I don't think it is totally incorrect to use "on"...
Besides, already have a lot of these print on other places, so if we are going to change it should be on a separate commit.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 9 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:990]: If there is no radio on the database, then the HAL will not be allocated in the first place. So that okay.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 12 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: The rest of the function should not be performed because it will be depended on the fact that we will get the radio struct from the database.
It should not happen obviously, and if it happens the son_slave will be stuck on some state trying to restart again and again, which is exactly what needs to be done, inorder it will be fixed.
Anyway I changed the print type tp `ERROR`

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 13 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:36]: Fixed.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 14 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:195]: Hmm maybe, anyway it is needed for not breaking the operationally of the code while moving code to the task on the next commits.

File: agent/src/beerocks/slave/tasks/topology_task.h (hunk 16 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/topology_task.h]: Fixed.

File: agent/src/beerocks/slave/tasks/topology_task.cpp (hunk 15 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/topology_task.cpp:38]: Fixed.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 17 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1145]: Actually it should have been `remained`. Anyway fixed.

File: agent/src/beerocks/slave/tasks/topology_task.cpp (hunk 18 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/topology_task.cpp:78]: It was like this before on the backhaul manager `after_select` function. It is the only thing the work function does. For me, it looks fine.

File: agent/src/beerocks/slave/tasks/topology_task.cpp (hunk 19 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/topology_task.cpp]: Fixed

File: agent/src/beerocks/slave/tasks/topology_task.cpp (hunk 20 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/topology_task.cpp:180]: What you've suggested is the same thing as what is written now.
Calling another function just for doing basic `find` on a container is redundant IMO.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 22 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:368]: It is describing what actually being done. It just happens on the Topology Task event handler.
And it is a good thing to print it here because it helps to understand what is the source of the Topology notification.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 23 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1821]: Same answer.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 24 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Fixed.

File: agent/src/beerocks/slave/tasks/topology_task.cpp (hunk 25 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/topology_task.cpp]: Fixed.

File: agent/src/beerocks/slave/tasks/topology_task.cpp (hunk 26 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/topology_task.cpp:218]: But then it would have break it...

File: ci/cppcheck/cppcheck_existing_issues.txt (hunk 27 above)
Comment (morantr) [reply ci/cppcheck/cppcheck_existing_issues.txt:11]: I prefer to fix it on the last commit.

File: agent/src/beerocks/slave/tasks/topology_task.cpp [hunk 28]
```diff
@@ -36,6 +36,53 @@ TopologyTask::TopologyTask(backhaul_manager &btl_ctx, ieee1905_1::CmduMessageTx
...
+        send_topology_discovery();
+    }
+
+    // Each platform must send Topology Discovery message every 60 seconds to its first circle
+    // 1905.1 neighbors.
+    // Iterate on each of known 1905.1 neighbors and check if we have received in the last
+    // 60 seconds a Topology Discovery message from it. If not, remove this neighbor from our list
+    // and send a Topology Notification message.
+    static constexpr uint8_t DISCOVERY_NEIGHBOUR_REMOVAL_TIMEOUT_SEC =
+        ieee1905_1_consts::DISCOVERY_NOTIFICATION_TIMEOUT_SEC + 3; // 3 seconds grace period.
```
Comment (itayx) [inline agent/src/beerocks/slave/tasks/topology_task.cpp:59]: nitpick
Add constexpr for the grace period

**Review by itayx — APPROVED** [review]
LGTM

File: agent/src/beerocks/slave/tasks/topology_task.cpp (hunk 28 above)
Comment (morantr) [reply agent/src/beerocks/slave/tasks/topology_task.cpp:59]: Not critical. There is a comment on that.
